File: sdk/go/osmp/brigade/base_helpers.go
// Brigade base helpers — opcode-existence check via active ASD.
//
// Faithful Go port of sdk/python/osmp/brigade/base_helpers.py.
package brigade

import "github.com/octid-io/cloudless-sky/sdk/go/osmp"

// OpcodeExists checks whether namespace:opcode is in the active ASD.
// Stations should consult this before proposing — emitting SAL with an
// opcode that doesn't exist in the loaded dictionary will fail validation
// AND signals to the receiver an action it cannot dispatch.
func OpcodeExists(namespace, opcode string) bool {
	ops, ok := osmp.ASDFloorBasis[namespace]
	if !ok {
		return false
	}
	_, exists := ops[opcode]
	return exists
}

// AllOpcodes returns all opcodes in a namespace from the active ASD.
func AllOpcodes(namespace string) []string {
	ops, ok := osmp.ASDFloorBasis[namespace]
	if !ok {
		return nil
	}
	out := make([]string, 0, len(ops))
	for op := range ops {
		out = append(out, op)
	}
	return out
}

File: sdk/python/osmp/brigade/base_helpers.py
"""Helpers shared by stations — opcode-existence check, etc."""
from __future__ import annotations

from ..protocol import ASD_BASIS


def opcode_exists(namespace: str, opcode: str) -> bool:
    """Check whether namespace:opcode is in the active ASD.

    Stations should consult this before proposing — emitting SAL with an
    opcode that doesn't exist in the loaded dictionary will fail validation
    AND signals to the receiver an action it cannot dispatch.
    """
    return opcode in ASD_BASIS.get(namespace, {})


def all_opcodes(namespace: str) -> list[str]:
    """Return all opcodes in a namespace from the active ASD."""
    return list(ASD_BASIS.get(namespace, {}).keys())
